components/brave_shields/browser/brave_shields_p3a.cc

```diff
@@ -71,10 +140,114 @@ void RecordShieldsFingerprintSetting(ControlType setting) {
+                                  : kDomainAdsSettingsBelowHistogramName;
+  int above_total = DomainCountRelativeToGlobalSetting(
+      profile_prefs, is_fingerprint, global_setting, true);
+  int below_total = DomainCountRelativeToGlobalSetting(
+      profile_prefs, is_fingerprint, global_setting, false);
+  VLOG(1) << "BraveShieldsP3A: Recording counts: is_fp=" << is_fingerprint
+          << " above=" << above_total << " below=" << below_total;
+  int above_bucket_val =
+      std::lower_bound(kDomainCountBuckets, std::end(kDomainCountBuckets),
+                       above_total) -
+      kDomainCountBuckets;
+  int below_bucket_val =
+      std::lower_bound(kDomainCountBuckets, std::end(kDomainCountBuckets),
+                       below_total) -
+      kDomainCountBuckets;
+  base::UmaHistogramExactLinear(above_hg_name, above_bucket_val,
+                                kDomainCountBucketCount);
+  base::UmaHistogramExactLinear(below_hg_name, below_bucket_val,
+                                kDomainCountBucketCount);
+}
+
+void RecordShieldsDomainSettingCountsWithChange(PrefService* profile_prefs,
+                                                bool is_fingerprint,
+                                                ControlType global_setting,
+                                                ControlType* prev_setting,
+                                                ControlType new_setting) {
+  if (profile_prefs == nullptr) {
+    return;
+  }
+  if (prev_setting != nullptr) {
+    UpdateDomainSettingCount(profile_prefs, is_fingerprint, *prev_setting, -1);
+    VLOG(1) << "BraveShieldsP3A: Decreasing prev setting count: prev_setting="
+            << *prev_setting << " is_fp=" << is_fingerprint << " count="
+            << GetDomainSettingCount(profile_prefs, is_fingerprint,
+                                     *prev_setting);
+  }
+  UpdateDomainSettingCount(profile_prefs, is_fingerprint, new_setting, 1);
+  VLOG(1) << "BraveShieldsP3A: Increasing new setting count: new_setting="
+          << new_setting << " is_fp=" << is_fingerprint << " count="
+          << GetDomainSettingCount(profile_prefs, is_fingerprint, new_setting);
+  RecordShieldsDomainSettingCounts(profile_prefs, is_fingerprint,
+                                   global_setting);
+}
+
+void MaybeRecordInitialShieldsSettings(PrefService* profile_prefs,
+                                       HostContentSettingsMap* map) {
+  if (profile_prefs->GetBoolean(kFirstReportedPrefName)) {
+    return;
+  }
+  VLOG(1) << "BraveShieldsP3A: Starting initial report for profile";
+
+  ControlType global_ads_setting = GetCosmeticFilteringControlType(map, GURL());
+  ControlType global_fp_setting = GetFingerprintingControlType(map, GURL());
+  RecordShieldsAdsSetting(global_ads_setting);
+  RecordShieldsFingerprintSetting(global_fp_setting);
+
+  ShieldsSettingCounts fp_counts = GetFPSettingCount(map);
+  ShieldsSettingCounts ads_counts = GetAdsSettingCount(map);
```

Comment:
When performing the first recording of the metrics for a given profile, the domain-specific settings are counted by iterating through the HostContentSettingsMap rules. This will ensure that any settings made by the user before the browser upgrade will be included in the counts.

Reply:
a comment in the code